Support for Active Directory multidomain forest. (#93)
matrix-org/matrix-synapse-ldap3: README.rst

        bind_password: "ch33kym0nk3y"
        #filter: "(objectClass=posixAccount)"

> 

Reviewer: I suggest a new section here:

```suggestion

Active Directory forest support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Active Directory forest accounts are also supported with custom
separator (`=` or `/`) in the following form: `login<separator>domain`.

Please note that common domain separators `@` and `\\` can not be used in Matrix

> User Indentifiers (https://matrix.org/docs/spec/appendices#user-identifiers)

Reviewer: Indentifiers -> Identifiers

> Active Directory forest accounts are also supported with custom
> separator (`=` or `/`) in the following form: `login<separator>domain`.
> 
> Please note that common domain separators `@` and `\\` can not be used in Matrix
> User Indentifiers (https://matrix.org/docs/spec/appendices#user-identifiers)
> and `/` separator is used by default.
> 
> Let's say you have several domains in example.com forest:

Reviewer: I think this could be reworded a bit.

```suggestion
If the ``ad_forest`` flag is set to ``true``, an Active Directory forest will be
searched for the login details.

In this mode, the user enters their login details in the form ``<login>/<domain>``. (The 
normal domain separators, ``@`` and ``\\``, cannot be used in Matrix User Identifiers, so 
``/`` is used by default instead.)
```

.. code:: yaml

   password_providers:
    - module: "ldap_auth_provider.LdapAuthProvider"
      config:
        enabled: true
        mode: "search"
        uri: "ldap://main.example.com:389"
        base: "dc=example,dc=com"
        # Must be true for this feature to work
        ad_forest: true
        # Optional. Users from this domain may login
        # without specifying domain part
        default_domain: main.example.com

>         # Optional. Default /
>         mxid_domain_separator: /

Reviewer: ```suggestion
        # Optional. Specifies the domain separator character that users should enter. Defaults to `/`.
        mxid_domain_separator: /
```

        attributes:
           # This must be set to userPrincipalName
           # when ad_forest is true
           uid: "userPrincipalName"
           mail: "mail"
           name: "givenName"
        bind_dn: "cn=hacker,ou=svcaccts,dc=example,dc=com"
        bind_password: "ch33kym0nk3y"
        filter: "(objectClass=user)"

> With this configuration you can use logins:

Reviewer: ```suggestion
With this configuration the user can log in with either ``someuser/main.example.com`` or ``someuser``.
```

etc

        bind_password: "ch33kym0nk3y"
        #filter: "(objectClass=posixAccount)"

Active Directory forest support
-------------------------------

If the ``active_directory`` flag is set to ``true``, an Active Directory forest will be
searched for the login details.
In this mode, the user enters their login details in one of the forms:

- ``<login>/<domain>``
- ``<domain>\<login>``

In either case, this will be mapped to the Matrix UID ``<login>/<domain>`` (The 
normal AD domain separators, ``@`` and ``\``, cannot be used in Matrix User Identifiers, so 
``/`` is used instead.)

Let's say you have several domains in the ``example.com`` forest:

.. code:: yaml

   password_providers:
    - module: "ldap_auth_provider.LdapAuthProvider"
      config:
        enabled: true
        mode: "search"
        uri: "ldap://main.example.com:389"
        base: "dc=example,dc=com"
        # Must be true for this feature to work
        active_directory: true
        # Optional. Users from this domain may log in without specifying the domain part
        default_domain: main.example.com
        attributes:

>            # This must be set to `userPrincipalName` when `active_directory` is true

Reviewer: this is now incorrect (or at least misleading)?

           uid: "userPrincipalName"
           mail: "mail"
           name: "givenName"
        bind_dn: "cn=hacker,ou=svcaccts,dc=example,dc=com"
        bind_password: "ch33kym0nk3y"

With this configuration the user can log in with either ``main.example.com\someuser``,
``someuser/main.example.com`` or ``someuser``.

Users of other domains in the ``example.com`` forest can log in with ``domain\login``
or ``login/domain``.

> Please note that ``userPrincipalName`` or similary looking LDAP attribute in the format

Reviewer: ```suggestion
Please note that ``userPrincipalName`` or a similar-looking LDAP attribute in the format
```

> ``login@domain`` must be used when ``active_directory`` option is enabled.

Reviewer: ```suggestion
``login@domain`` must be used when the ``active_directory`` option is enabled.
```
